### ISSUE TITLE
chore(buffers): add unit test for modifying a data file when the buffer is still open (disk v2)

### ISSUE THREAD

**Issue by tobz** [vectordotdev/vector#10898]
In #10868, it was called out that I still have a TODO for writing a test which exercises the behavior of the buffer when a data file is modified while the buffer is still open, as our buffer initialization process is the only place where we calculate our buffer size to ensure correctness.  This would exercise what happens when a data file is, say, partially written, and writes keep proceeding.

**Comment by jszwedko** [comment]
Should we add this to https://github.com/vectordotdev/vector/issues/9476?

**Comment by tobz** [comment]
@jszwedko Just added it. 👍🏻

**Comment by tobz** [comment]
So, after thinking it through and considering some recent changes, I'm going to close this issue, and I'll explain why it doesn't matter anymore.

The original code comment that prompted this issue was as follows:
>  // If the writer has marked themselves as done, and the buffer has been emptied, then
            // we're done and can return.  We have to look at something besides simply the writer
            // being marked as done to know if we're actually done or not, and "buffer size" is better
            // than "total records" because we update buffer size when handling acknowledgements,
            // whether it's an individual ack or an entire file being deleted.
            //
            // If we used "total records", we could end up stuck in cases where we skipped
            // corrupted records, but hadn't yet had a "good" record that we could read, since the
            // "we skipped records due to corruption" logic requires performing valid read to
            // detect, and calculate a valid delta from.
            //
            // TODO: Validate this with a test that messes with a data file without having to reload
            // the buffer, since reloading the buffer will recalculate our buffer size.

Essentially, in the core logic of the reader's `next` method, which tries to read a record, we eventually have to figure out if the file we're reading from has no more data because the writer had to roll over to the next file, or because it was actually the last record written, and the writer has closed itself, thus implying we too are also done reading.

Originally, this was tricky because we hadn't yet simplified our use of the ledger data to harden against spurious errors.  We were checking front and back, looking at multiple values, all in service of trying to figure out if we were "done", and if the numbers we used to calculate that we were done were actually _valid_.

As part of an [in-progress branch](https://github.com/vectordotdev/vector/tree/tobz/disk-buffer-v2-traitify-io) to use property-based testing against the disk buffer v2 implementation, we've made a crucial change which is that the writer does not update the size of the buffer, or the "next record ID to use" field, until a `flush` successfully occurs.

What this means is that our logic in the reader, which currently looks something like:

```
if writer_marked_themselves_done {
    let unread_bytes_in_buffer = ...
    let unread_records_in_buffer = ...;

    if unread_bytes_in_buffer == 0 || unread_records_in_buffer == 0 {
        return Ok(None);
    }
}
```

will currently -- without the aforementioned branch -- potentially get stuck waiting for bytes that the writer thought it wrote, but never did.   There's a few reasons this could happen, some of which aren't even really bugs but simply the writer API not being resistant enough to misuse.  Essentially, we always expect the writing task to call `flush` after doing a write, or writes... if they never do that, the data _may_ have been implicitly flushed if we were lucky enough to overflow the buffered writer's internal buffers...  but otherwise, the writer just assumed the flush was coming, and preemptively updated the ledger.  

The invariant we're able to provide with the new behavior in the aforementioned branch is that we _never_ update_ the ledger until we successfully `flush`, which implies flushing all buffered writes out to the file.  It does not, however, imply actually doing an `fsync` to disk, because we ultimately rely on the operating system to provide cache coherency within its page/buffer cache implementation.  In practice, this means the reader should correctly observe all writes that the OS reported as successfully processing _as long as_ the writer also then successfully flushes those writes to the file.

We now come to why we do not need this test.  If we depend on the invariant that a successful flush within the writer is reported correctly, then we should never end up in a situation where the buffer size/record count is non-zero unless the writer successfully flushed, and if it successfully flushed, then the data should be readable by the reader.

Any violation of these invariants would be a violation of the operating system's responsibility to ensure page/buffer cache coherency, and would represent an unrecoverable error, not unlike a disk failing and returning an error during `fsync`, and the process, or at the very least, the component using the buffer, should immediately exit and reload so that the buffer can be initialized and recovered to the furthest possible extent.

Said another way, a data file whose contents became incoherent between the writer and reader process, where there was not an outright process or system crash, is an error we cannot possibly hope to recover from at runtime.
